Add implementation of manager-wide ListUnits

alexlarsson opened the pull request:
This calls ListUnits on each online node and collects the responses, adding the name first in the struct.

The free_func_t type requires us to add -Wno-cast-function-type to cflags, because we need to be able to cast real destroy functions to the generic one.

Signed-off-by: Alexander Larsson <alexl@redhat.com>

alexlarsson committed:
Add implementation of manager-wide ListUnits

This calls ListUnits on each online node and collects the responses,
adding the name first in the struct.

The free_func_t type requires us to add -Wno-cast-function-type to
cflags, because we need to be able to cast real destroy functions to
the generic one.

Signed-off-by: Alexander Larsson <alexl@redhat.com>

sdunnagan commented on src/agent/agent.c:
```diff
@@ -233,7 +233,7 @@ static int list_units_callback(sd_bus_message *m, void *userdata, UNUSED sd_bus_
         }
 
-        sd_bus_message *reply = NULL;
+        _cleanup_sd_bus_message_ sd_bus_message *reply = NULL;
```
Cool!  I see that this is needed because of TAKE_PTR() used in message_new_reply() in
systemd/src/libsystemd/sd-bus/bus-message.c

alexlarsson replied:
Well, sort of. I mean, independent of how this is implemented internally in sd-bus it returns an object with a ref count, and it needs to be freed eventually.

sdunnagan commented on src/manager/manager.c:
```diff
@@ -252,9 +254,191 @@ static int manager_method_ping(sd_bus_message *m, UNUSED void *userdata, UNUSED
+                sd_bus_message *m;
+                AgentRequest *agent_req;
+        } sub[0];
+} ListUnitRequest;
```
This type could be renamed ```ListUnitsRequest```.

sdunnagan commented on src/manager/manager.c:
```diff
@@ -252,9 +254,191 @@ static int manager_method_ping(sd_bus_message *m, UNUSED void *userdata, UNUSED
+
+
+static void manager_method_list_units_done(ListUnitRequest *req) {
+        /* All sub-requests are done, collect results and freee when done */
```
free

sdunnagan commented on src/manager/manager.c:
```diff
@@ -252,9 +254,191 @@ static int manager_method_ping(sd_bus_message *m, UNUSED void *userdata, UNUSED
+        }
+}
+
+static void manager_method_list_units_maybe_done(ListUnitRequest *req) {
```
Maybe rename this ```manager_method_check_if_list_units_done()```.

alexlarsson replied:
I think that primarily makes it longer, not necessary easier to understand.

sdunnagan commented on src/manager/manager.c:
```diff
@@ -252,9 +254,191 @@ static int manager_method_ping(sd_bus_message *m, UNUSED void *userdata, UNUSED
+        ListUnitRequest *req = NULL;
+        Node *node = NULL;
+
+        req = malloc0(sizeof(*req) + sizeof(req->sub[0]) * manager->n_nodes);
```
Seems like we need to check ```n_nodes``` here in case it's some huge number.

alexlarsson replied:
n_nodes is the number of nodes that the sysadmin configured for the cluster though. Why would that be huge?

alexlarsson replied:
It wouldn't hurt if we added a helper function for allocations like this though, that handled overflows in the computations.

alexlarsson replied:
I added malloc0_array() that does the required overflow checks.

sdunnagan commented on src/manager/manager.c:
```diff
@@ -252,9 +254,191 @@ static int manager_method_ping(sd_bus_message *m, UNUSED void *userdata, UNUSED
+                Node *node;
+                sd_bus_message *m;
+                AgentRequest *agent_req;
+        } sub[0];
```
How about renaming this ```sub_req```?

sdunnagan commented on src/manager/manager.c:
```diff
@@ -252,9 +254,191 @@ static int manager_method_ping(sd_bus_message *m, UNUSED void *userdata, UNUSED
+
+        int i = 0;
+        LIST_FOREACH(nodes, node, manager->nodes) {
+                _cleanup_agent_request_ AgentRequest *agent_req = node_request_list_units(
```
I believe this ```_cleanup_agent_request_``` can be removed because the pointer to AgentRequest was already stolen inside ```node_request_list_units()``` and the memory is freed later in ```list_unit_request_free()```.

alexlarsson replied:
It all depends on your point of view. Neither of the two is wrong. The ownership of a ref is transferred from the return of node_request_list_units() into the local variable agent_req, and then transferred ot the sub[].agent_req location (via pointer_steal). 

Since this *always* happens we could leave out the cleanup of the agent_req to make the code simpler. On the other hand, it would be less explicit about how we track ownership, and if any code changes later that makes the assignment of agent_req not always happen, then we would need to add this back. For example, if we introduce another call site between the two that could fail with an error.

sdunnagan replied:
Cool, makes sense!

alexlarsson committed:
common: Add malloc0_array()

This allows you to easily allocate a base + N element aray without
worrying about overflows of the computations.

Signed-off-by: Alexander Larsson <alexl@redhat.com>
